controllers/flowcollector_controller.go

```diff
@@ -90,271 +89,105 @@ func NewFlowCollectorReconciler(client client.Client, scheme *runtime.Scheme, co
+	r.status.SetReady()
+	defer r.status.Commit(ctx, r.Client)
+
+	err := r.reconcile(ctx)
 	if err != nil {
-		log.Error(err, "Failed to get FlowCollector")
-		return ctrl.Result{}, err
-	} else if desired == nil {
-		return ctrl.Result{}, nil
-	}
-
-	ns := getNamespaceName(desired)
-	if err := cleanup.CleanPastReferences(ctx, r.Client, ns); err != nil {
+		l.Error(err, "FlowCollector reconcile failure")
+		// Set status failure unless it was already set
+		if !r.status.HasFailure() {
+			r.status.SetFailure("FlowCollectorGenericError", err.Error())
+		}
 		return ctrl.Result{}, err
 	}
-	r.watcher.Reset(ns)
 
-	var didChange, isInProgress bool
-	previousNamespace := desired.Status.Namespace
+	return ctrl.Result{}, nil
+}
 
-	// obtain default cluster ID - api is specific to openshift
-	if r.permissions.Vendor(ctx) == discover.VendorOpenShift && globals.DefaultClusterID == "" {
-		cversion := &configv1.ClusterVersion{}
-		key := client.ObjectKey{Name: "version"}
-		if err := r.Client.Get(ctx, key, cversion); err != nil {
-			log.Error(err, "unable to obtain cluster ID")
-		} else {
-			globals.DefaultClusterID = cversion.Spec.ClusterID
-		}
+func (r *FlowCollectorReconciler) reconcile(ctx context.Context) error {
+	clh, desired, err := helper.NewFlowCollectorClientHelper(ctx, r.Client)
+	if err != nil {
+		return fmt.Errorf("failed to get FlowCollector: %w", err)
+	} else if desired == nil {
+		return nil
 	}
 
+	ns := helper.GetNamespace(&desired.Spec)
+	previousNamespace := r.status.GetDeployedNamespace(desired)
 	loki := helper.NewLokiConfig(&desired.Spec.Loki)
-	reconcilersInfo := r.newCommonInfo(ctx, desired, ns, previousNamespace, &loki, func(b bool) { didChange = b }, func(b bool) { isInProgress = b })
+	reconcilersInfo := r.newCommonInfo(clh, ns, previousNamespace, &loki)
 
-	err = r.reconcileOperator(ctx, &reconcilersInfo, desired)
-	if err != nil {
-		return ctrl.Result{}, err
+	if ret, err := r.checkFinalizer(ctx, desired, &reconcilersInfo); ret {
+		return err
 	}
 
+	if err := cleanup.CleanPastReferences(ctx, r.Client, ns); err != nil {
+		return err
+	}
+	r.watcher.Reset(ns)
+
 	// Create reconcilers
-	flpReconciler := flowlogspipeline.NewReconciler(&reconcilersInfo, r.config.FlowlogsPipelineImage)
 	var cpReconciler consoleplugin.CPReconciler
-	if r.availableAPIs.HasConsolePlugin() {
-		cpReconciler = consoleplugin.NewReconciler(&reconcilersInfo, r.config.ConsolePluginImage)
+	if r.mgr.HasConsolePlugin() {
+		cpReconciler = consoleplugin.NewReconciler(reconcilersInfo.NewInstance(r.mgr.Config.ConsolePluginImage, r.status))
 	}
 
 	// Check namespace changed
 	if ns != previousNamespace {
-		if err := r.handleNamespaceChanged(ctx, previousNamespace, ns, desired, &flpReconciler, &cpReconciler); err != nil {
-			return ctrl.Result{}, r.failure(ctx, conditions.CannotCreateNamespace(err), desired)
+		if previousNamespace != "" && r.mgr.HasConsolePlugin() {
```

Comment:
Not related to this PR.

Handling namespace change create some complexity that is not necessary IMO.

Apparently it is possible to make a field immutable with kubebuilder pattern. I think it could be a nice compromise.
What do you think?

Reply:
I agree that the namespace thing is a pain.. We can make it immutable yes - we then loose the ability to reconfigure it on the fly, but I guess it's quite a rare operation anyway, and the workaround would just be to manually delete then re-install the flowcollector at the desired namespace. That would interrupt flow collection, but it's also interrupted when you modify namespace, so.. yeah I think I'm good with that proposal :-)